Upgrade tool for 2.8

rohitsinha54 opened the pull request:
An initial PR for Upgrade tool for review.
The Upgrade tool currently does:
- Creates the default namespace
- Does application meta upgrade
- Does log.meta upgrade
- Does Archive/File upgrades - The required ones like streams, logs etc. We do not move program archive as moving them does not make sense since a recompiled app needs to be deployed which will create them.

rohitsinha54 commented on cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrade.java:
```diff
@@ -0,0 +1,195 @@
+  private final Transactional<UpgradeTable, Table> appMDS;
+
+  public MDSUpgrade() {
+    this.appMDS = Transactional.of(executorFactory, new Supplier<UpgradeTable>() {
```
Can someone suggest a better way of accessing app.meta table here if there's one ?
Thanks.

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataFabricDistributedModule.java:
```diff
@@ -68,7 +68,7 @@ public void configure() {
    */
   @Singleton
-  private static final class ThriftClientProviderSupplier implements Provider<ThriftClientProvider> {
+  public static final class ThriftClientProviderSupplier implements Provider<ThriftClientProvider> {
```
Why does this need to be public?

rohitsinha54 replied:
Because I want to use this to bind it my in the injector in the Main class.

albertshau replied:
how come we can't use this module in the Main class?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+/**
+ * Abstract class for Upgrade
+ */
+public abstract class AbstractUpgrade {
```
Better name - ``AbstractUpgrader``? Nouns for class names.

anew replied:
Is the intention that this should implement the Upgrade interface?

rohitsinha54 replied:
No. I just want to have all the shared variables and function among different modules in here.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  /**
+   * Sets up a {@link DatasetFramework} instance for standalone usage.  NOTE: should NOT be used by applications!!!
```
What does this comment mean? Or is it just carried over from somewhere else?

rohitsinha54 replied:
Carried over.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+    DatasetFramework datasetFramework =
+      new NamespacedDatasetFramework(new InMemoryDatasetFramework(registryFactory),
+                                     new DefaultDatasetNamespace(cConf));
+    // TODO: this doesn't sound right. find out why its needed.
```
Could you please remove this comment? I think I added it but forgot to remove. This is needed.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+                                     new DefaultDatasetNamespace(cConf));
+    // TODO: this doesn't sound right. find out why its needed.
+    datasetFramework.addModule(Id.DatasetModule.from(Constants.SYSTEM_NAMESPACE, "orderedTable"),
+                               new HBaseOrderedTableModule());
```
``HBaseOrderedTableModule()`` is going away, in #1236 

That PR is ready to merge. I can work with you to apply its changes to this PR.

rohitsinha54 replied:
I resolved the conflict.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  /**
+   * Creates a non-namespaced {@link DatasetFramework} to access existing datasets which are not namespaced
+   */
+  public DatasetFramework createNonNamespaceDSFramework(Injector injector) throws DatasetManagementException {
```
Which datasets are namespaced, and which ones are not?

rohitsinha54 replied:
Lets discuss this in person. Its for the problem we saw where the table was created like cdap.system.cdap.system.table.name since the other framework is namespaced.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+   * @param validPrefixes the valid prefixes
+   * @return boolean which is true if the key start with validPrefixes else false
+   */
+  boolean checkKeyValidality(String key, String[] validPrefixes) {
```
Just ``checkKeyValidity``.

Although can we make it more specific? valid with respect to?

chtyim replied:
Better call `validateKey`? check and validate are kind of redundant.

chtyim replied:
Or with common Java naming convention, it can be `isKeyValid`.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramController.java:
```diff
@@ -100,6 +100,20 @@ private State(ProgramRunStatus runStatus) {
+     * @param status : the status
+     * @return the state for the status or null if there is no defined state for the given status
+     */
+    public static State getControllerStateByStatus (ProgramRunStatus status) {
```
better to put this in the ProgramController.State enum.  Also, this is a very weird method, there are many states that map to ProgramRunStatus.RUNNING.  This one will return STARTING all the time. Is that what you want?

albertshau replied:
also remove extra space between ```ByStatus (ProgramRunStatus```

rohitsinha54 replied:
I think it should be fine because the program completed status is written when the program is completed which will map to STOPPED(ProgramRunStatus.COMPLETED) or ERROR(ProgramRunStatus.FAILED) but if we look at the above case if it give us     STARTING(ProgramRunStatus.RUNNING) it will get mapped to RUNNING when it gets written which is what others maps too. So I think it should be fine.

albertshau replied:
I'm not sure I understand.  How is this being used?  I don't think we should have a method like this because it is completely misleading.  RUNNING does NOT always mean STARTING, which is what this method returns.

chtyim replied:
Why do you need this method during upgrade? Shouldn't the upgrade just copy the existing values into different rows?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  static final Gson GSON;
+
+  static {
+    GsonBuilder builder = new GsonBuilder();
```
you can do this in one line and avoid the static block.

rohitsinha54 replied:
Talked in person. Ignoring this.

anwar6953 replied:
Its because of this: `ApplicationSpecificationAdapter.addTypeAdapters`?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  /**
+   * Sets up a {@link DatasetFramework} instance for standalone usage.  NOTE: should NOT be used by applications!!!
+   */
+  public static DatasetFramework createRegisteredDatasetFramework(Injector injector)
```
why does this need to be static?

albertshau replied:
doesn't seem like this or ```createNonNamespaced``` should be inside this class, as the frameworks should be created once by the tool and then passed in to all the upgraders.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  static final String CDAP_WITH_FORWARD_SLASH = Constants.Logging.SYSTEM_NAME + FORWARD_SLASH;
+  static final String DEVELOPER_STRING = "developer";
+
+  static DatasetFramework namespacedFramework;
```
why are these all static?

albertshau replied:
also, seems like everything that is package private should be protected or private.

albertshau replied:
also confused about how they get set. seems like most should be final and set in a constructor

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrade.java:
```diff
@@ -0,0 +1,56 @@
+   * Upgrades the streams archive path
+   */
+  private void upgradeStreamFilepath() throws IOException {
+    List<Location> streamLocations = locationFactory.create(EMPTY_STRING).append(FORWARD_SLASH +
```
is this different than just doing the create without the empty string?

bdmogal replied:
yup, should just be ``locationFactory.create("streams").list()``

bdmogal replied:
Also, better to not use the constant ``Constants.Service.STREAMS`` here. It seems like a service name, and could change in future?

chtyim replied:
Never ever use string concat with slash to form the path. Always do `locationFactory.create(basePath).append(subdir).append(file)`.

chtyim replied:
The stream location should be from `CConfiguration` (`locationFactory.create(cConf.get(Constants.Stream.BASE_DIR))`).

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+    return nonNamespacedFramework;
+  }
+
+  Location renameLocation(String oldLocation, String newLocation) throws URISyntaxException, IOException {
```
seems like these should be protected

chtyim replied:
Who call this? I don't see any usage of this method

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrade.java:
```diff
@@ -0,0 +1,56 @@
+                                                      oldLocation.getName());
+      LOG.info("Upgrading Stream archive for {}", oldLocation);
+      renameLocation(oldLocation, newLocation);
+    }
```
this should remove the old streams directory once it's done

anwar6953 replied:
also, why move the stream dirs one at a time? 
Can't move the entire /streams dir into default/streams?

rohitsinha54 replied:
Done.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrade.java:
```diff
@@ -0,0 +1,119 @@
+    // Upgrade all datasets in system namespace
+    Id.Namespace systemNamespace = Id.Namespace.from(Constants.SYSTEM_NAMESPACE);
+    for (DatasetSpecification spec : framework.getInstances(systemNamespace)) {
+      System.out.println(String.format("Upgrading dataset: %s, spec: %s", spec.getName(), spec.toString()));
```
log instead of print

albertshau replied:
or if we're printing, it should print everywhere instead of log some places print others.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrade.java:
```diff
@@ -0,0 +1,119 @@
+
+  private static void upgradeUserTables(final Injector injector) throws Exception {
+    // We assume that all tables in USER namespace belong to OrderedTable type datasets. So we loop thru them
+    // and upgrading with the help of HBaseOrderedTableAdmin
```
this is not a valid assumption, is there more work incoming?

bdmogal replied:
I think this comment was also carried over from the old tool. We should definitely change this if its possible to have other types.

rohitsinha54 replied:
Is it possible to have different Type ? I think so. Isn't it ?

albertshau replied:
I think I read this as all datasets instead of all tables so you can ignore. Though all Tables are now Tables (OrderedTable was removed by Gary's changes I believe).

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrade.java:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+
+  private static void upgradeUserTables(final Injector injector) throws Exception {
```
shouldn't be static, and shouldnt' pass in an injector.  The injector should really only be used in one class to get all objects needed, and those objects should be passed to the correct classes

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -158,115 +162,28 @@ private Action parseAction(String action) {
-    // Upgrade all user hbase tables
-    upgradeUserTables(injector);
+    LOG.debug("Upgrading System and User Datasets ...");
+    Upgrade dsUpgrade = new DatasetUpgrade();
```
as mentioned before, don't pass in the injector.  These classes should take the required objects in from their constructors, and the constructors should be private and annotated with ```@Inject```.  Then here you use the injector to get DatasetUpgrade, MDSUpgrade, etc

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Upgrade.java:
```diff
@@ -0,0 +1,26 @@
+ * An interface to define Upgrade. Modules which needs to be upgraded should implement this.
+ */
+public interface Upgrade {
+  void upgrade(Injector injector) throws Exception;
```
shouldn't take any args. Also, AbstractUpgrade should implement this.  Or you can just move the upgrade method to the abstract class and remove this interface.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+        Scanner rows = ctx.get().scan(null, null);
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(Arrays.copyOfRange(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
```
toString() has a version with offset and length, you don't need to use Arrays.copyOfRange()

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+          String key = Bytes.toString(Arrays.copyOfRange(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
+                                                         row.getRow().length)).trim();
+          Map<byte[], byte[]> allColumns = row.getColumns();
+          for (byte[] startTime : allColumns.keySet()) {
```
nit: a little better to use allColumns.entrySet()

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+              fileMetaDataManager.writeMetaData(newKey, startTimeMs, locationFactory.create(new URI(newPath)));
+              LOG.info("Upgrading logs for {}", key);
+            } else if (!checkKeyValidality(key, OTHERS)) {
+              LOG.warn("Unknown key {} found. Failed to Upgrade", key);
```
what happens in the scenario? Does the tool stop? Is it impossible to upgrade once we're here?

rohitsinha54 replied:
I think we can continue with upgrade and things might or might now work depending on the key and where it is required. I will get rid of the RuntimeException so that the upgrade proceeds.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+              newPath = upgradePath(key, oldPath);
+              newKey = upgradeKey(key);
+              renameLocation(new URI(oldPath).resolve("."), new URI(newPath).resolve("."));
+              fileMetaDataManager.writeMetaData(newKey, startTimeMs, locationFactory.create(new URI(newPath)));
```
this and the previous line are not atomic. We should be sure to handle the case where rename location goes through but this does not. Perhaps check the paths before actually renaming?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+    String newLocation = null;
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      if (location != null) {
+        newLocation = location.substring(0, location.lastIndexOf(CDAP_WITH_FORWARD_SLASH)) +
```
this is confusing to read, can you add a comment with an example of what this does

chtyim replied:
Can you operate on `Location` object instead of `String`?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrade.java:
```diff
@@ -0,0 +1,195 @@
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(row.getRow()).trim();
+          if (key.contains(AppMetadataStore.TYPE_APP_META)) {
```
not sure what the keys look like, but is using ```contains``` safe? Are there any user provided values (such as app name) in the key?

albertshau replied:
similarly, what happens if we stop halfway through? Will this be able to identify what has already been migrated and skip over those?

bdmogal replied:
Should use ``MDSKey`` for deserialization, get the string and then compare with equals, I think. Much cleaner and safer. Perhaps create the ``MDSKey`` here and pass it to the handler methods below, instead of creating it in each method. The String contains operations you have here may not be as safe.

bdmogal replied:
I mean pass the splitter after skipping the record type, or something on similar lines.

rohitsinha54 replied:
@bdmogal : Awesomeness. :+1: 
Thanks.

albertshau commented:
took a pass, most comments are around restructuring the way things are constructed and around making sure we can re-run the tool if it fails midway through an upgrade.  Next time please try to get the framework in first, then open smaller PRs for upgrading individual components

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrade.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Class to Upgrade the archive directories
+ */
+public class ArchiveUpgrade extends AbstractUpgrade implements Upgrade {
```
Why does it ``extends`` and ``implements``?

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrade.java:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Upgrades the streams archive path
+   */
+  private void upgradeStreamFilepath() throws IOException {
```
Just `upgradeStream`. Potentially it can be more than file path.

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+ */
+public abstract class AbstractUpgrade {
+
+  public static final String EMPTY_STRING = "";
```
Why need a constant?

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  public static final String EMPTY_STRING = "";
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractUpgrade.class);
+  static final byte[] COLUMN = Bytes.toBytes("c");
+  static final String FORWARD_SLASH = "/";
```
Don't do this if you use it to construct file path. Also use the `Location.append` method.

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrade.java:
```diff
@@ -0,0 +1,155 @@
+  Location renameLocation(Location oldLocation, Location newLocation) throws IOException {
+    if (!newLocation.exists() && oldLocation.exists()) {
+      newLocation.mkdirs();
+      return oldLocation.renameTo(newLocation);
```
If the rename failed with exception, you should remove the `newLocation`.

chtyim replied:
BTW, why need to `mkdir` before the rename? If you worry the parent doesn't exists, you should call `Locations.getParent(newLocation).mkdirs()`.

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(Arrays.copyOfRange(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
+                                                         row.getRow().length)).trim();
```
Why need to trim?

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java:
```diff
@@ -0,0 +1,148 @@
+            if (key.startsWith(Constants.Logging.SYSTEM_NAME) || key.startsWith(DEVELOPER_STRING)) {
+              newPath = upgradePath(key, oldPath);
+              newKey = upgradeKey(key);
+              renameLocation(new URI(oldPath).resolve("."), new URI(newPath).resolve("."));
```
Why need to call `resolve`?

chtyim commented on cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrade.java:
```diff
@@ -0,0 +1,195 @@
+                                     FORWARD_SLASH + "archive"));
+  }
+
+  static final class UpgradeTable implements Iterable<Table> {
```
private class.

anew commented:
@rohitsinha54 it would have been nice to send a PR for just the framework, and then a separate PR for each upgrade step.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrade.java:
```diff
@@ -0,0 +1,195 @@
+   * @return the new archive location with namespace
+   */
+  private String updateAppArchiveLocation(String appId, String archiveLocation) {
+    return archiveLocation.replace("programs/archive" + FORWARD_SLASH + DEVELOPER_STRING,
```
Can we use ``String.format`` to construct the two Strings/CharSequences here separately and then call ``replace``? Very hard to read this way.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeSetup.java:
```diff
@@ -0,0 +1,103 @@
+import org.apache.twill.zookeeper.ZKClientService;
+
+/**
+ * Does the necessary initial setup before Upgrading difference Services and Modules
```
nit: 'U' shouldn't be upper-case, also s/difference/different/

rohitsinha54 commented:
@bdmogal @albertshau : I have addressed all the comments. Please have another look. I have went through the comments 2-3 times to make sure I addressed them but if in case I have missed something. I will be glad if you guys can point it out again.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:
```diff
@@ -57,12 +57,12 @@
+  public static final String TYPE_STREAM = "stream";
+  public static final String TYPE_RUN_RECORD_STARTED = "runRecordStarted";
+  public static final String TYPE_RUN_RECORD_COMPLETED = "runRecordCompleted";
   private static final String TYPE_PROGRAM_ARGS = "programArgs";
```
So we don't need to upgrade ``programArgs`` keys?

rohitsinha54 replied:
Do we need to ? The Upgrade doc does not mention anything about it and I didn't see them in the hbase table too. I will look into it. It might be a good idea to keep the upgrade doc requirements updated if we find new things as we proceed.

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/hbase/AbstractHBaseDataSetAdmin.java:
```diff
@@ -190,7 +190,7 @@ public static void setVersion(HTableDescriptor tableDescriptor) {
   public static ProjectInfo.Version getVersion(HTableDescriptor tableDescriptor) {
     String value = tableDescriptor.getValue(CDAP_VERSION);
-    return value == null ? null : new ProjectInfo.Version(value);
+    return new ProjectInfo.Version(value);
```
why this change? ``ProjectInfo.Version`` handles nulls?

rohitsinha54 replied:
This was like this before. @abaranau changed it to return null except of 0 if the table does not have a version and that breaks thing so I discussed with him and we decided to revert it back.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrader.java:
```diff
@@ -0,0 +1,97 @@
+public abstract class AbstractUpgrader {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractUpgrader.class);
+  protected static final byte[] COLUMN = Bytes.toBytes("c");
+  protected static final String DEVELOPER_STRING = "developer";
```
Call it ``DEVELOPER_ACCOUNT``? The data type already states that its a ``String``?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrader.java:
```diff
@@ -0,0 +1,97 @@
+   * @return new location if and only if the file or directory is successfully moved; null otherwise.
+   * @throws IOException
+   */
+  protected Location renameLocation(Location oldLocation, Location newLocation) throws IOException {
```
annotate with ``@Nullable``

On a side note, is it better to have this method's contract to either return a ``Location`` (which could be null), or throw an ``Exception``, but not both?

rohitsinha54 replied:
We return the return value of underlying all to renameTo which can be null. Also the rename can throw IOException which we want to catch to delete the newLocation and propagate. 
I marked it with ``@Nullable``

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrader.java:
```diff
@@ -0,0 +1,87 @@
+    Location oldLocation = locationFactory.create(cConf.get(Constants.Stream.BASE_DIR));
+    Location newLocation = Locations.getParent(oldLocation).append(Constants.DEFAULT_NAMESPACE)
+      .append(cConf.get(Constants.Stream.BASE_DIR));
+    if (renameLocation(oldLocation, newLocation) != null) {
```
what happens if this rename fails (meaning this condition is not met)?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrader.java:
```diff
@@ -0,0 +1,87 @@
+    Location newLocation2 = locationFactory.create(Constants.DEFAULT_NAMESPACE).append("logs").append("avro");
+
+    // move to temp location
+    renameLocation(oldLogLocation1, tempLocation1);
```
why make this a 2-step process? rename should be atomic, no? or do you do it because the source/destination directories are under the same directory and there may be problems because of that?

albertshau replied:
+1.  Doing it the current way also means you cannot run this again if it fails part way through.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/ArchiveUpgrader.java:
```diff
@@ -0,0 +1,87 @@
+   */
+  private void upgradeLogs() throws IOException {
+    // move log files under cdap and developer dir
+    Location oldLogLocation1 = locationFactory.create("logs").append("avro").append(Constants.Logging.SYSTEM_NAME);
```
There is a constant called ``Constants.Logging.LOG_BASE_DIR`` or something for this.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -0,0 +1,140 @@
+
+    // Upgrade all datasets in system namespace
+    Id.Namespace systemNamespace = Id.Namespace.from(Constants.SYSTEM_NAMESPACE);
+    for (DatasetSpecification spec : framework.getInstances(systemNamespace)) {
```
I think there may be a problem here. The ``DatasetFramework.getInstances(Id.Namespace)`` method was added recently, and it probably depends on namespacing in datasets, which is not available in older versions. Can you please check?

rohitsinha54 replied:
I will have a look but the history shows that this line was modified by you recently. Did you checked when you modified it ?

rohitsinha54 replied:
Discussed in person. @bdmogal will do the changes in a different PR.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -0,0 +1,140 @@
+  private void upgradeSystemDatasets(DatasetFramework framework) throws Exception {
+
+    // Upgrade all datasets in system namespace
+    Id.Namespace systemNamespace = Id.Namespace.from(Constants.SYSTEM_NAMESPACE);
```
Use ``Constants.SYSTEM_NAMESPACE_ID``

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -0,0 +1,140 @@
+
+    for (HTableDescriptor desc : hAdmin.listTables()) {
+      String tableName = desc.getNameAsString();
+      TableId tableId = TableId.from(tableName);
```
Can you add a TODO here that this will not work for upgrade from 2.8 to later versions? We will have to come up with a better way of mapping table name to dataset name and vice versa?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+   * @return the log {@link Location}
+   * @throws IOException
+   */
+  private Location updateLogLocation(Location location, String nameSpace) throws IOException {
```
nit: no need for upper-casing 'S'

albertshau replied:
shouldn't we be moving the log files? Or is that happening somewhere else?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+    String logFilename = location.getName();
+    Location parentLocation = Locations.getParent(location);  // strip filename
+    String date = parentLocation.getName();
+    parentLocation = Locations.getParent(parentLocation); // strip date
```
Each of these ``Locations.getParent(parentLocation)`` may return a null, theoretically. However, it should never happen since this is the expected path from the log.meta table. May be worth precondition'ing them.

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrader.java:
```diff
@@ -0,0 +1,208 @@
+            runRecordCompletedHandler(row, keyParts);
+          } else if (!isKeyValid(key, OTHERS)) {
+            LOG.warn("Invalid Metadata with key {} found in {}", key, DefaultStore.APP_META_TABLE);
+            throw new RuntimeException("Invalid metadata with key " + key);
```
``IllegalStateException``?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrader.java:
```diff
@@ -0,0 +1,208 @@
+    String pId = keyParts.getString();
+
+    writeTempRunRecordStart(appId, programId, pId, startTs);
+    store.setStop(Id.Program.from(Constants.DEFAULT_NAMESPACE, appId, programId), pId, runRecord.getStopTs(),
```
this cleans up the temp record written above, right?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -84,7 +112,105 @@ private String getDescription() {
+      });
+    txService = injector.getInstance(TransactionService.class);
+    zkClientService = injector.getInstance(ZKClientService.class);
+//    locationFactory = injector.getInstance(LocationFactory.class);
```
remove if not needed

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -84,7 +112,105 @@ private String getDescription() {
+    cConf = CConfiguration.create();
+    hConf = HBaseConfiguration.create();
+
+    Injector injector = Guice.createInjector(
```
is it better to set up the injector in a separate ``init()`` method?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -157,119 +266,107 @@ private Action parseAction(String action) {
+    dsUpgrade.upgrade();
+
+    LOG.debug("Upgrading app meta data ...");
+    MDSUpgrader mdsUpgrad = injector.getInstance(MDSUpgrader.class);
```
typo - mdsUpgrader?

bdmogal commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -157,119 +266,107 @@ private Action parseAction(String action) {
-    // Upgrade system datasets
-    upgradeSystemDatasets(injector);
+  private void performUpgrade() throws Exception {
+    LOG.debug("Upgrading System and User Datasets ...");
```
May be a good idea to have these logs at ``INFO`` level, since this is a tool that people may want to monitor progress closely?

I would make them INFO atleast until we finish testing.

bdmogal commented:
Made one pass through it. Will do more. Looking much better than before - we're getting there.

bdmogal commented:
Also, you may want to check with @abaranau and @gsps1 to make sure that the changes they're making in the upgrade tool are included in your PR. There is some moving around of code here, so will help to have them take a look, although git should catch conflicts.

bdmogal commented:
Another thing to consider is how this can be extended to cover upgrades from other versions - e.g. 2.7 -> 2.8. 2.8 -> 3.0, etc, can be done in a separate PR, but something to think about.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrader.java:
```diff
@@ -0,0 +1,97 @@
+      } catch (Exception e) {
+        newLocation.delete();
+        LOG.warn("Failed to rename {} with {}", oldLocation, newLocation);
+        throw Throwables.propagate(e);
```
should just throw the exception you got instead of wrapping it in a runtime exception.  It's always an IOException right?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -0,0 +1,140 @@
+    upgradeSystemDatasets(namespacedFramework);
+
+    // Upgrade all user hbase tables
+    upgradeUserTables();
```
users can also have fileset datasets.  Need an upgrade path for those (can do it in a separate PR).  Is there a jira open for that?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+  private final Transactional<DatasetContext<Table>, Table> logMDS;
+  private final CConfiguration cConf;
+  private final TransactionSystemClient txClient;
+  private final DatasetFramework nonNamespaedFramework;
```
nit: typo. Namespaed -> Namespaced

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+    Location newLocation = null;
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      if (location != null) {
+        // Example path of this type: hdfs://cdap28-161119-1000.dev.continuuity.
```
never put actual hostnames in the code, even as comments

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+    Location location = locationFactory.create(new URI(oldLocation));
+    Location newLocation = null;
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      if (location != null) {
```
can it be null?  What does that mean, the old location doesn't exist?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+  /**
+   * Creates a new path depending on the old log file path
+   *
+   * @param key         the key for this log meta entry
```
remove extra spaces

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+              newKey = upgradeKey(key);
+              try {
+                fileMetaDataManager.writeMetaData(newKey, startTimeMs, newPath);
+                LOG.info("Upgrading logs for {}", key);
```
this log should either come before we upgrade, or it should say "Upgraded" instead of "Upgrading"

albertshau replied:
or both

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+                LOG.warn("No updating log file path in log meta table");
+                throw Throwables.propagate(t);
+              }
+            } else if (!isKeyValid(key, OTHERS)) {
```
how is this possible?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+                fileMetaDataManager.writeMetaData(newKey, startTimeMs, newPath);
+                LOG.info("Upgrading logs for {}", key);
+              } catch (Throwable t) {
+                LOG.warn("No updating log file path in log meta table");
```
"Failed updating ..." instead of "No updating ..."

albertshau replied:
log the exception too

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+                LOG.info("Upgrading logs for {}", key);
+              } catch (Throwable t) {
+                LOG.warn("No updating log file path in log meta table");
+                throw Throwables.propagate(t);
```
better just to throw the exception instead of wrapping in a runtime exception.

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java:
```diff
@@ -0,0 +1,203 @@
+              newPath = upgradePath(key, oldPath);
+              newKey = upgradeKey(key);
+              try {
+                fileMetaDataManager.writeMetaData(newKey, startTimeMs, newPath);
```
Should we delete the old entry?  What happens if we fail part way through. Next time it runs it will try to upgrade the same paths.  Is that handled correctly?

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -84,7 +112,105 @@ private String getDescription() {
+          bind(QueueAdmin.class).to(HBaseQueueAdmin.class).in(Singleton.class);
+          install(new TransactionModules().getDistributedModules());
+          bind(DatasetFramework.class).to(RemoteDatasetFramework.class);
+          new DataSetServiceModules().getDistributedModule();
```
this line isn't doing anything

albertshau commented on cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java:
```diff
@@ -84,7 +112,105 @@ private String getDescription() {
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          bind(HBaseTableUtil.class).toProvider(HBaseTableUtilFactory.class);
```
I'm not sure how all our guice modules are structured, but how come we have to do all these custom bindings?  Existing modules don't have what you need?

albertshau commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java:
```diff
@@ -87,13 +87,25 @@ public TransactionExecutor createExecutor(Iterable<TransactionAware> txAwares) {
+   * @param startTimeMs start log time associated with the file.
+   * @param location log file.
+   */
+  public void writeMetaData(final String logPartition,
```
this seems bad.  Doesn't look like this class intended to expose that the log partition is used in this way... is there any way to get the LoggingContext from the row key?

albertshau commented:
The framework looks pretty good.  Most of my comments were on the individual upgrade pieces, especially around failure conditions and making sure it is idempotent.

chtyim commented:
It seems like the upgrade tools requires quite a number of class to expose constants/methods as public, which I think in general is very bad, since it's a direct violation of good OO design (encapsulation). Why the internal logic needs to be exposed? (e.g. when you expose a constant and use it outside of the class to do row key matching, you have assumption on the internal of the class).

albertshau commented:
I think we have to split this up.  There are too many orthogonal pieces and comments in this PR.  Let's make one for just the framework which seems close to being able to merge.  Then we can have branches for each upgrade component, and we can parallelize the work if needed.
